File: backend/app/schemas/clinical.py
"""
臨床記錄 Schemas
  VitalSign / SoapNote / SoapDiagnosis / NursingNote
"""
from datetime import datetime
from typing import Optional

from pydantic import BaseModel


# ── VitalSign ─────────────────────────────────────────────────

class VitalSignCreate(BaseModel):
    weight_kg: Optional[float] = None
    temperature_c: Optional[float] = None
    heart_rate_bpm: Optional[int] = None
    respiratory_rate_bpm: Optional[int] = None
    systolic_bp_mmhg: Optional[int] = None
    capillary_refill_sec: Optional[float] = None
    body_condition_score: Optional[int] = None  # 1–9


class VitalSignRead(BaseModel):
    id: int
    visit_id: int
    weight_kg: Optional[float] = None
    temperature_c: Optional[float] = None
    heart_rate_bpm: Optional[int] = None
    respiratory_rate_bpm: Optional[int] = None
    systolic_bp_mmhg: Optional[int] = None
    capillary_refill_sec: Optional[float] = None
    body_condition_score: Optional[int] = None
    mucous_membrane_color_id: Optional[int] = None
    mucous_membrane_color_name: Optional[str] = None
    is_superseded: bool
    created_at: datetime
    created_by_name: Optional[str] = None

    model_config = {"from_attributes": True}


# ── SoapDiagnosis ─────────────────────────────────────────────

class SoapDiagnosisCreate(BaseModel):
    free_text: str


class SoapDiagnosisRead(BaseModel):
    id: int
    free_text: Optional[str] = None
    is_superseded: bool
    created_at: datetime
    created_by_name: Optional[str] = None

    model_config = {"from_attributes": True}


# ── SoapNote ──────────────────────────────────────────────────

class SoapNoteCreate(BaseModel):
    subjective: Optional[str] = None
    objective: Optional[str] = None
    assessment: Optional[str] = None
    plan: Optional[str] = None
    diagnoses: list[SoapDiagnosisCreate] = []


class SoapNoteRead(BaseModel):
    id: int
    visit_id: int
    subjective: Optional[str] = None
    objective: Optional[str] = None
    assessment: Optional[str] = None
    plan: Optional[str] = None
    diagnoses: list[SoapDiagnosisRead] = []
    is_superseded: bool
    created_at: datetime
    created_by_name: Optional[str] = None

    model_config = {"from_attributes": True}


# ── ClinicalSummary（看板卡片用） ────────────────────────────────

class ClinicalSummary(BaseModel):
    """最新一筆生命徵象 + 主診斷，供看板卡片快速顯示"""
    latest_weight_kg: Optional[float] = None
    latest_temperature_c: Optional[float] = None
    latest_heart_rate_bpm: Optional[int] = None
    latest_diagnosis: Optional[str] = None  # 最新 SoapNote 的主診斷 free_text
    has_pending_lab: bool = False            # 是否有待結果的檢驗單


# ── NursingNote ───────────────────────────────────────────────

class NursingNoteCreate(BaseModel):
    note_text: str


class NursingNoteRead(BaseModel):
    id: int
    visit_id: int
    note_text: str
    is_superseded: bool
    created_at: datetime

    model_config = {"from_attributes": True}


# ── LabOrder ──────────────────────────────────────────────────


class LabResultItemCreate(BaseModel):
    analyte_id: int
    value_numeric: Optional[float] = None
    value_text: Optional[str] = None
    is_abnormal: Optional[bool] = None
    notes: Optional[str] = None


class LabResultItemRead(BaseModel):
    id: int
    analyte_id: int
    analyte_name: str       # JOIN 帶入
    unit: Optional[str] = None
    analyte_type: str
    value_numeric: Optional[float] = None
    value_text: Optional[str] = None
    is_abnormal: Optional[bool] = None
    notes: Optional[str] = None
    is_superseded: bool
    created_at: datetime
    created_by_name: Optional[str] = None

    model_config = {"from_attributes": True}


class LabOrderCreate(BaseModel):
    test_type_id: int
    notes: Optional[str] = None


class LabOrderRead(BaseModel):
    id: int
    visit_id: int
    test_type_id: int
    test_type_name: str     # JOIN 帶入
    status: str             # pending | resulted | cancelled
    notes: Optional[str] = None
    resulted_at: Optional[datetime] = None
    resulted_by_name: Optional[str] = None
    is_superseded: bool
    created_at: datetime
    created_by_name: Optional[str] = None
    result_items: list[LabResultItemRead] = []

    model_config = {"from_attributes": True}


class LabResultSubmit(BaseModel):
    """技術員填寫完所有指標後一次提交"""
    items: list[LabResultItemCreate]

File: frontend/src/api/clinical.ts
import api from "@/api";

// ── 型別定義 ──────────────────────────────────────────────────

export interface VitalSignCreate {
  weight_kg?: number | null;
  temperature_c?: number | null;
  heart_rate_bpm?: number | null;
  respiratory_rate_bpm?: number | null;
  systolic_bp_mmhg?: number | null;
  capillary_refill_sec?: number | null;
  body_condition_score?: number | null;
}

export interface VitalSignRead extends VitalSignCreate {
  id: number;
  visit_id: number;
  is_superseded: boolean;
  created_at: string;
  created_by_name: string | null;
}

export interface SoapDiagnosisCreate {
  free_text: string;
  is_primary?: boolean;
}

export interface SoapDiagnosisRead {
  id: number;
  free_text: string | null;
  is_primary: boolean;
  is_superseded: boolean;
  created_at: string;
}

export interface SoapNoteCreate {
  subjective?: string | null;
  objective?: string | null;
  assessment?: string | null;
  plan?: string | null;
  diagnoses?: SoapDiagnosisCreate[];
}

export interface SoapNoteRead {
  id: number;
  visit_id: number;
  subjective: string | null;
  objective: string | null;
  assessment: string | null;
  plan: string | null;
  diagnoses: SoapDiagnosisRead[];
  is_superseded: boolean;
  created_at: string;
}

export interface NursingNoteCreate {
  note_text: string;
}

export interface NursingNoteRead {
  id: number;
  visit_id: number;
  note_text: string;
  is_superseded: boolean;
  created_at: string;
}

export interface ClinicalSummary {
  latest_weight_kg: number | null;
  latest_temperature_c: number | null;
  latest_heart_rate_bpm: number | null;
  latest_diagnosis: string | null;
}

// ── API 函式 ──────────────────────────────────────────────────

export const clinicalApi = {
  // Vital Signs
  getVitalSigns: (visitId: number): Promise<VitalSignRead[]> =>
    api.get(`/visits/${visitId}/vital-signs`).then((r) => r.data),

  createVitalSign: (
    visitId: number,
    body: VitalSignCreate
  ): Promise<VitalSignRead> =>
    api.post(`/visits/${visitId}/vital-signs`, body).then((r) => r.data),

  // SOAP Notes
  getSoapNotes: (visitId: number): Promise<SoapNoteRead[]> =>
    api.get(`/visits/${visitId}/soap-notes`).then((r) => r.data),

  createSoapNote: (
    visitId: number,
    body: SoapNoteCreate
  ): Promise<SoapNoteRead> =>
    api.post(`/visits/${visitId}/soap-notes`, body).then((r) => r.data),

  // Nursing Notes
  getNursingNotes: (visitId: number): Promise<NursingNoteRead[]> =>
    api.get(`/visits/${visitId}/nursing-notes`).then((r) => r.data),

  createNursingNote: (
    visitId: number,
    body: NursingNoteCreate
  ): Promise<NursingNoteRead> =>
    api.post(`/visits/${visitId}/nursing-notes`, body).then((r) => r.data),

  getClinicalSummary: (visitId: number): Promise<ClinicalSummary> =>
    api.get(`/visits/${visitId}/clinical-summary`).then((r) => r.data),
};
